arch/loongarch/kernel/idle.c

```diff
@@ -4,13 +4,15 @@
  *
  * Copyright (C) 2020-2022 Loongson Technology Corporation Limited
  */
+#include "asm-generic/irq_pipeline.h"
 #include <linux/cpu.h>
 #include <linux/irqflags.h>
 #include <asm/cpu.h>
 #include <asm/idle.h>
 
 void __cpuidle arch_cpu_idle(void)
 {
+	hard_local_irq_enable();
```

Comment:
Does this try to replace `raw_local_irq_enable` with `hard_local_irq_enable` originally?

Reply:
According to comments in line 233 at `kernel/sche/idle.c` linux6.6 after patch dovetail, it's trying to catch mishandling of the CPU's interrupt disable flag.This pattern can also find in 5.15 dovetail patch line 4365.Only difference between older version and newer version is one line of additional if statement check for do `hard_local_irq_enable`  exactly when `hard_irqs_disabled() == true` but it shouldn't.
5.15 dovetail patch arm64
 ```c
--- a/arch/arm64/kernel/idle.c
+++ b/arch/arm64/kernel/idle.c
@@ -42,5 +42,6 @@ void noinstr arch_cpu_idle(void)
 	 * tricks
 	 */
 	cpu_do_idle();
+	hard_cond_local_irq_enable();
 	raw_local_irq_enable();
 }

```

newer version dovetail `kernel/sched/idle.c`
```c
exit_idle:
	__current_set_polling();

	/*
	 *  Catch mishandling of the CPU's interrupt disable flag when
	 *  pipelining IRQs.
	 */
	if (WARN_ON_ONCE(irq_pipeline_debug() && hard_irqs_disabled()))
		hard_local_irq_enable();

	/*
	 * It is up to the idle functions to reenable local interrupts
	 */
	if (WARN_ON_ONCE(irqs_disabled()))
		local_irq_enable();

```